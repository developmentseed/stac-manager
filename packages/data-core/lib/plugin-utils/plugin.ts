/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars */

import { SchemaField, SchemaFieldObject } from '../schema/types';

export type PluginEditSchema =
  | (SchemaFieldObject & { type: 'root' })
  | undefined
  | symbol;

export interface PluginHook {
  // Name of the target plugin.
  name: string;

  //  The `onAfterInit` hook is executed after the target plugin's `init`
  //  function.
  onAfterInit?: (pluginInstance: Plugin, data: any) => void;

  // The `onAfterEditSchema` hook is composed with the target plugin's
  // `editSchema` function.
  onAfterEditSchema?: (
    pluginInstance: Plugin,
    formData: any,
    schema: PluginEditSchema
  ) => PluginEditSchema;
}

const HIDDEN: unique symbol = Symbol('hidden');
const HOOKS: unique symbol = Symbol('hooks');

export abstract class Plugin {
  static readonly HIDDEN: typeof HIDDEN = HIDDEN;
  static readonly HOOKS: typeof HOOKS = HOOKS;

  [HOOKS]: PluginHook[] = [];

  name: string = 'Plugin';

  init(data: any) {}

  editSchema(formData?: any): PluginEditSchema {
    return undefined;
  }

  enterData(data: any): Record<string, any> {
    throw new Error(`Plugin [${this.name}] must implement enterData`);
  }

  exitData(data: any): Record<string, any> {
    throw new Error(`Plugin [${this.name}] must implement exitData`);
  }

  /**
   * Registers a hook to be applied on a given plugin.
   *
   * @param targetName - The name of the target plugin to which the hook will be
   * applied.
   * @param hooks - The hook details.
   */
  registerHook(targetName: string, hooks: Omit<PluginHook, 'name'>) {
    this[Plugin.HOOKS].push({ name: targetName, ...hooks });
  }
}

export type PluginConfigResolved = Plugin | Plugin[] | undefined | null;

export type PluginConfigResolver = (data: any) => PluginConfigResolved;

export type PluginConfigItem = PluginConfigResolved | PluginConfigResolver;
