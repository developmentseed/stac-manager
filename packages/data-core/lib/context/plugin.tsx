import React, { createContext, useContext } from 'react';

import { Plugin } from '../plugin-utils/plugin';

// Plugin context provider
const PluginContext = createContext<Plugin | undefined>(undefined);

/**
 * Context provider for each plugin configuration
 * 
 * @param props.config Plugin configuration object
 * @param props.children Child components
 */
export const PluginProvider = (props: {
  plugin: Plugin;
  children: React.ReactNode;
}) => {
  return (
    <PluginContext.Provider value={props.plugin}>
      {props.children}
    </PluginContext.Provider>
  );
};

export const usePlugin = () => {
  const plugin = useContext(PluginContext);

  if (!plugin) {
    throw new Error('usePlugin must be used within a PluginProvider');
  }

  return plugin;
};
