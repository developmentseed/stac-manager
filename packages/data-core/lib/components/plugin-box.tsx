import React from 'react';
import { Code } from '@chakra-ui/react';
import { useFormContext, useWatch } from 'react-hook-form';

import { PluginProvider } from '../context/plugin';
import { Plugin } from '../plugin-utils/plugin';
import { ErrorBox } from './error-box';
import { SchemaField } from '../schema/types';

interface PluginBoxProps {
  plugin: Plugin;
  children: ({ field }: { field: SchemaField }) => JSX.Element;
}

/**
 * Prepares the plugin schema and sets up the plugin context.
 * Provides the plugin schema to the children function.
 * 
 * @param props.plugin The plugin to render
 * @param props.children The children function to render the plugin schema.
 */
export function PluginBox(props: PluginBoxProps) {
  const { plugin, children } = props;

  const { getValues } = useFormContext();
  const editSchema = plugin.editSchema(getValues());

  useWatch({
    name: plugin.watchFields || [],
    disabled: !plugin.watchFields?.length
  });

  if (!editSchema) {
    return (
      <ErrorBox>
        Plugin <Code color='red'>{plugin.name}</Code> has no edit schema.
      </ErrorBox>
    );
  }

  if (editSchema === Plugin.HIDDEN) {
    return null;
  }

  return (
    <PluginProvider plugin={plugin}>
      {children({ field: editSchema as SchemaField })}
    </PluginProvider>
  );
}
