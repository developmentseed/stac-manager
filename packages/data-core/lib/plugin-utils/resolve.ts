import { cloneDeep } from 'lodash-es';
import { Plugin, PluginConfigItem } from './plugin';

/**
 * Resolves the plugin config.
 *
 * @param plugins - An array of plugin configuration items which can be
 * instances of `Plugin` or functions that return a `Plugin`.
 * @param data - The data to be passed to plugin functions if they are not
 * instances of `Plugin`.
 * @returns An array of processed `Plugin` instances after applying hooks.
 */
export const resolvePlugins = (plugins: PluginConfigItem[], data: any) => {
  const p = plugins
    .flatMap((pl) => {
      if (pl instanceof Plugin) {
        return pl;
      } else if (typeof pl === 'function') {
        return pl(data);
      }
      return;
    })
    .filter((p) => p instanceof Plugin);

  return applyHooks(p);
};

/**
 * Applies hooks from the provided plugins to their respective targets.
 *
 * This function iterates over each plugin and applies hooks such as
 * `onAfterInit` and `onAfterEditSchema` to the corresponding target plugins.
 * The hooks are executed in the context of the source plugin.
 *
 * @param plugins - List of plugins. All the source and target plugins must be
 * on the list.
 *
 *
 * @remarks
 * - The `onAfterInit` hook is executed after the target plugin's `init`
 *   function.
 * - The `onAfterEditSchema` hook is composed with the target plugin's
 *   `editSchema` function.
 *
 * @example
 * ```typescript
 * class MyPlugin extends Plugin {
 *  name = 'MyPlugin';
 *
 *  [Plugin.HOOKS]: [
 *   {
 *     name: 'pluginA', // Target plugin
 *     onAfterInit: async (targetInstance, data) => { },  // Executes after pluginA's init function.
 *     onAfterEditSchema: (targetInstance, formData, origEditSchema) => { } // Composes with pluginA's editSchema function and returns a new one.
 *   },
 *   {
 *     name: 'pluginB', // Target plugin
 *     onAfterInit: async (targetInstance, data) => { }, // Executes after pluginB's init function.
 *   }
 *  ];
 * }
 *
 * applyHooks(plugins);
 * ```
 */
export const applyHooks = (plugins: Plugin[]) => {
  const pluginsCopy = cloneDeep(plugins);

  for (const plSource of pluginsCopy) {
    for (const hook of plSource[Plugin.HOOKS]) {
      // Target where to apply the hook
      const plTarget = plugins.find((p) => p.name === hook.name);
      if (!plTarget) {
        continue;
      }

      // The onAfterInit hook is made by executing one function after another.
      if (hook.onAfterInit) {
        const fn = hook.onAfterInit;
        const origInit = plTarget.init;
        plTarget.init = async (data: any) => {
          await origInit.call(plTarget, data);
          await fn.call(plSource, plTarget, data);
        };
      }

      // The onAfterEditSchema hook is made by composing functions.
      if (hook.onAfterEditSchema) {
        const fn = hook.onAfterEditSchema;
        const origEditSchema = plTarget.editSchema;
        plTarget.editSchema = (formData?: any) => {
          return fn.call(
            plSource,
            plTarget,
            formData,
            origEditSchema.call(plTarget, formData)
          );
        };
      }
    }
  }

  return pluginsCopy;
};
