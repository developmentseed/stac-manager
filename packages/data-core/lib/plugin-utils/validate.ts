import * as Yup from 'yup';
import { yupToFormErrors } from 'formik';

import {
  SchemaFieldArray,
  SchemaFieldJson,
  SchemaFieldNumber,
  SchemaFieldObject,
  SchemaFieldString
} from '../schema/types';
import { Plugin } from './plugin';

/**
 * Validates the data against the edit schemas defined by the plugins.
 * @param {Plugin[]} plugins - The list of plugins.
 * @param {any} data - The data to validate.
 * @returns {[any, any]} - Returns a tuple with the validated data or errors.
 */
export function validatePluginsFieldsData(plugins: Plugin[], data: any) {
  const validationSchema = plugins.reduce((acc, plugin) => {
    const editSchema = plugin.editSchema(data);
    if (!editSchema || editSchema === Plugin.HIDDEN) return acc;

    const schema = createObjectSchema({
      ...(editSchema as SchemaFieldObject),
      type: 'object'
    });

    return acc.concat(schema);
  }, Yup.object());

  try {
    const result = validationSchema.validateSync(data, { abortEarly: false });
    return [result, null];
  } catch (error: any) {
    const e = yupToFormErrors(error);
    return [null, e];
  }
}

/**
 * Creates a Yup object schema based on the provided data.
 * @param {SchemaFieldObject} data - The schema field object.
 * @param {boolean} [isRequired] - Whether the schema is required.
 * @returns {Yup.ObjectSchema<object>} - The Yup object schema.
 */
function createObjectSchema(
  data: SchemaFieldObject,
  isRequired?: boolean
): Yup.ObjectSchema<object, Yup.AnyObject, object, ''> {
  const required = data.required || [];
  const properties = data.properties || {};

  const mappedProperties = Object.keys(properties).reduce((acc, key) => {
    const field = properties[key];
    const type = field.type;
    const isFieldRequired = required.includes(key);

    let schema;
    if (type === 'object' || type === 'root') {
      schema = createObjectSchema(field, isFieldRequired);
    } else if (type === 'string') {
      schema = createStringSchema(field, isFieldRequired);
    } else if (type === 'number') {
      schema = createNumberSchema(field, isFieldRequired);
    } else if (type === 'json') {
      schema = createJsonSchema(field, isFieldRequired);
    } else if (type === 'array') {
      schema = createArraySchema(field, isFieldRequired);
    } else {
      throw new Error('Invalid item type');
    }

    return {
      ...acc,
      [key]: schema
    };
  }, {});

  if (isRequired) {
    return Yup.object(mappedProperties).required();
  } else {
    return Yup.object(mappedProperties);
  }
}

/**
 * Creates a Yup string schema based on the provided data.
 * @param {SchemaFieldString} data - The schema field string.
 * @param {boolean} [isRequired] - Whether the schema is required.
 * @returns {Yup.StringSchema} - The Yup string schema.
 */
function createStringSchema(data: SchemaFieldString, isRequired?: boolean) {
  let schema = Yup.string().label(
    typeof data.label === 'string' ? data.label : 'Value'
  );

  if (data.enum) {
    schema = schema.test('enum', '${path} value is invalid', (value) => {
      if (value === '' && !isRequired) return true;

      const inOption = data.enum.some(([v]) => v === value);
      if (inOption) {
        return true;
      }

      if (data.allowOther) {
        return typeof value === data.allowOther.type;
      }
      return false;
    });
  }
  return isRequired ? schema.required() : schema;
}

/**
 * Creates a Yup number schema based on the provided data.
 * @param {SchemaFieldNumber} data - The schema field number.
 * @param {boolean} [isRequired] - Whether the schema is required.
 * @returns {Yup.NumberSchema} - The Yup number schema.
 */
function createNumberSchema(data: SchemaFieldNumber, isRequired?: boolean) {
  const schema = Yup.number()
    .label(typeof data.label === 'string' ? data.label : 'Value')
    .typeError('Value must be a number');

  return isRequired ? schema.required() : schema;
}

/**
 * Creates a Yup JSON schema based on the provided data.
 * @param {SchemaFieldJson} data - The schema field JSON.
 * @param {boolean} [isRequired] - Whether the schema is required.
 * @returns {Yup.ObjectSchema} - The Yup JSON schema.
 */
function createJsonSchema(data: SchemaFieldJson, isRequired?: boolean) {
  const schema = Yup.object().label(
    typeof data.label === 'string' ? data.label : 'Value'
  );

  return isRequired ? schema.required() : schema;
}

/**
 * Creates a Yup array schema based on the provided data.
 * @param {SchemaFieldArray} data - The schema field array.
 * @param {boolean} [isRequired] - Whether the schema is required.
 * @returns {Yup.ArraySchema} - The Yup array schema.
 */
function createArraySchema(
  data: SchemaFieldArray,
  isRequired?: boolean
): Yup.ArraySchema<any[] | undefined, Yup.AnyObject, '', ''> {
  const field = data.items || {};
  const type = field.type;

  let itemSchema;
  if (type === 'object') {
    itemSchema = createObjectSchema(field);
  } else if (type === 'string') {
    itemSchema = createStringSchema(field);
  } else if (type === 'number') {
    itemSchema = createNumberSchema(field);
  } else if (type === 'json') {
    itemSchema = createJsonSchema(field);
  } else if (type === 'array') {
    itemSchema = createArraySchema(field);
  } else {
    throw new Error('Invalid item type');
  }

  const schema = Yup.array()
    .label(typeof data.label === 'string' ? data.label : 'Value')
    .of(itemSchema)
    .min(data.minItems || 0)
    .max(data.maxItems || Infinity);

  return isRequired ? schema.required() : schema;
}
