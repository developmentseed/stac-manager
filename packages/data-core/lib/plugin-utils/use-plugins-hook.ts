import { useCallback, useEffect, useMemo, useState } from 'react';
import { defaultsDeep } from 'lodash-es';

import { Plugin, PluginConfigItem } from './plugin';
import { resolvePlugins } from './resolve';
import { usePluginConfig } from '../context/plugin-config';
import { FormDataStructure, schemaToDataStructure } from '../schema';

type UsePluginsHook =
  | {
      isLoading: true;
      plugins: undefined;
      formData: undefined;
      toOutData: (formData: any) => undefined;
    }
  | {
      isLoading: false;
      plugins: Plugin[];
      formData: any;
      toOutData: (formData: any) => any;
    };

const usePlugins = (plugins: PluginConfigItem[], data: any): UsePluginsHook => {
  const [readyPlugins, setReadyPlugins] = useState<Plugin[]>();

  useEffect(() => {
    async function load() {
      // if (!data) return;

      const resolvedPlugins = resolvePlugins(plugins, data);
      await Promise.all(resolvedPlugins.map((pl) => pl.init(data)));
      setReadyPlugins(resolvedPlugins);
    }
    load();
  }, [plugins, data]);

  const formData = useMemo(() => {
    if (!readyPlugins) return;

    const emptyStructure = readyPlugins.reduce<FormDataStructure>(
      (acc, pl: Plugin) => {
        const schema = pl.editSchema();
        if (!schema || typeof schema === 'symbol') return acc;

        return {
          ...acc,
          ...(schemaToDataStructure(schema) as FormDataStructure)
        };
      },
      {}
    );

    return readyPlugins.reduce(
      (acc: any, pl: Plugin) => defaultsDeep(acc, pl.enterData(data)),
      emptyStructure
    );
  }, [readyPlugins]);

  const toOutData = useCallback(
    (formData: any) =>
      readyPlugins &&
      readyPlugins.reduce(
        (acc: any, pl: Plugin) => ({
          ...acc,
          ...pl.exitData(formData)
        }),
        {}
      ),
    [readyPlugins]
  );

  if (!readyPlugins) {
    return {
      isLoading: true,
      plugins: undefined,
      formData: undefined,
      /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
      toOutData: (formData: any) => undefined
    };
  }

  return {
    isLoading: false,
    plugins: readyPlugins,
    formData,
    toOutData
  };
};

export function useCollectionPlugins(data: any) {
  const config = usePluginConfig();

  return usePlugins(config.collectionPlugins, data);
}

export function useItemPlugins(data: any) {
  const config = usePluginConfig();

  return usePlugins(config.itemPlugins, data);
}
