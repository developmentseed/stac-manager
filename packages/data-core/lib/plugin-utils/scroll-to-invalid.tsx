import { useEffect } from 'react';
import { useFormikContext } from 'formik';
import { getFirstPath } from './validate';

/**
 * Scrolls to the first invalid field according to the formik errors.
 */
export function ScrollToInvalidField() {
  const { errors, isSubmitting, isValidating } = useFormikContext();

  useEffect(() => {
    // Run on submission after validation took place.
    if (!errors || !isSubmitting || isValidating) return;

    const path = getFirstPath(errors);
    // Use the parent node because if the element is a hidden input, it won't
    // scroll.
    const element = document.querySelector(`[name="${path}"]`)
      ?.parentNode as HTMLElement;
    element?.scrollIntoView({ block: 'center', behavior: 'smooth' });
  }, [errors, isSubmitting, isValidating]);

  return null;
}
