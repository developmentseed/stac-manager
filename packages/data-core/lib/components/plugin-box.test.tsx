import React from 'react';
import { render } from '@testing-library/react';
import { Formik } from 'formik';
import { ChakraProvider } from '@chakra-ui/react';

import { PluginBox } from './plugin-box';
import { Plugin } from '../plugin-utils/plugin';

const mockPlugin = {
  name: 'TestPlugin',
  editSchema: jest.fn()
};

// Custom renderer to add context or providers if needed
const renderWithProviders = (
  ui: React.ReactNode,
  { renderOptions = {} } = {}
) => {
  // You can wrap the component with any context providers here
  return render(ui, {
    wrapper: ({ children }) => (
      <ChakraProvider>
        <Formik initialValues={{}} onSubmit={() => {}}>
          {children}
        </Formik>
      </ChakraProvider>
    ),
    ...renderOptions
  });
};

describe('PluginBox', () => {
  it('renders ErrorBox when editSchema is null', () => {
    mockPlugin.editSchema.mockReturnValue(null);

    const { getByTestId } = renderWithProviders(
      <Formik initialValues={{}} onSubmit={() => {}}>
        <PluginBox plugin={mockPlugin as any}>
          {({ field }) => <div>{field.label}</div>}
        </PluginBox>
      </Formik>
    );

    expect(getByTestId('plugin-box-error')).toMatchSnapshot();
  });

  it('renders nothing when editSchema is Plugin.HIDDEN', () => {
    mockPlugin.editSchema.mockReturnValue(Plugin.HIDDEN);

    const { container } = render(
      <Formik initialValues={{}} onSubmit={() => {}}>
        <PluginBox plugin={mockPlugin as any}>
          {({ field }) => <div>{field.label}</div>}
        </PluginBox>
      </Formik>
    );

    expect(container.firstChild).toBeNull();
  });

  it('renders children when editSchema is valid', () => {
    const mockSchema = { type: 'string', label: 'testField' };
    mockPlugin.editSchema.mockReturnValue(mockSchema);

    const { getByText } = renderWithProviders(
      <Formik initialValues={{}} onSubmit={() => {}}>
        <PluginBox plugin={mockPlugin as any}>
          {({ field }) => <div>{field.label}</div>}
        </PluginBox>
      </Formik>
    );

    expect(getByText(/testField/i)).toBeInTheDocument();
  });
});
