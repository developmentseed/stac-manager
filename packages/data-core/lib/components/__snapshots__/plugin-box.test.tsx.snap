// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginBox renders ErrorBox when editSchema is null 1`] = `
<div
  class="css-133ygyc"
  data-testid="plugin-box-error"
>
  Plugin 
  <code
    class="chakra-code css-1fu8sp0"
  >
    TestPlugin
  </code>
   has no edit schema.
</div>
`;
